### PR TITLE
Defmt support for enums and structs

### PR DESCRIPTION
**Pull request by haata**
- Hidden behind defmt-impl feature
  (similar to how heapless handles defmt)

### CODE REVIEW

**Comment by stappersg** [comment]
But why?


Longer version:
The pull request has been seen, not yet seen by me is what it improves. Consider that it is me who isn't seen the larger picture. Please elaborate _defmt_ and tell what it makes better.

**Comment by haata** [comment]
USB tends to be timing sensitive, so in order to add logging inside a USB hal crate (what I'm currently working on https://github.com/atsam-rs/atsam4-hal/pull/49/files) defmt is necessary.

My current development flow is to use probe-run + defmt for logging and bracktrace support.
defmt (https://ferrous-systems.com/blog/defmt/ is a good read, but I'll give the tl;dr here) strips out debug messages from the firmware, places it on the host side, then uses a protobuf-like mechanism to pass any necessary arguments back to the host to reassemble the print message. The benefit, significantly smaller firmware images (if you have lots of messages) as well as faster logging as fewer bytes need to be sent over the debug link (e.g. RTT).

If these derive macros aren't added, I would need to maintain either a fork or wrap each enum and struct with another struct in order to handle log message formatting.

Does that make sense?

**Comment by stappersg** [comment]
On Tue, Aug 24, 2021 at 12:45:19AM -0700, Jacob Alexander wrote:
> USB tends to be timing sensitive, so in order to add logging inside a USB hal crate  defmt is necessary.
> 
>    ... further good text ... 
>  ...(https://ferrous-systems.com/blog/defmt/ is a good read ...
> 
> Does that make sense?


Yes, it does.


It's worth to add to the commit message.

**Comment by haata** [comment]
Sorry for the delay, I've been under the weather.

File: Cargo.toml [hunk 1]
```diff
@@ -24,6 +27,9 @@ control-buffer-256 = []
 # speed device, but the descriptors will be invalid.
 test-class-high-speed = []
 
+# Add defmt Format support for enums and structs
+defmt-impl = ["defmt"]
```
Comment (ryan-summers) [inline Cargo.toml:31]: Let's copy how `smoltcp` does this. I don't think we need to add an explicit feature flag. The user can just do `features=["defmt"]` and that should be sufficient.
https://github.com/smoltcp-rs/smoltcp/blob/cae682e0885502cc3fa4e122d44f67167addb095/Cargo.toml#L24

https://github.com/smoltcp-rs/smoltcp/blob/cae682e0885502cc3fa4e122d44f67167addb095/src/socket/icmp.rs#L25

**Review by ryan-summers — CHANGES_REQUESTED** [review]
Can you please also add a CHANGELOG entry for adding defmt implementations for various structures?

**Comment by haata** [comment]
Done

File: src/bus.rs [hunk 2]
```diff
@@ -281,6 +281,7 @@ impl<B: UsbBus> UsbBusAllocator<B> {
 
 /// A handle for a USB interface that contains its number.
 #[derive(Copy, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "defmt-impl", derive(defmt::Format))]
```
Comment (ryan-summers) [inline src/bus.rs:284]: All of the code blocks use feature `defmt-impl` instead of just `defmt`

**Review by ryan-summers — CHANGES_REQUESTED** [review]

File: src/bus.rs (hunk 2 above)
Comment (haata) [reply src/bus.rs]: Added another cargo check with the defmt feature.

Comment (haata) [reply src/bus.rs]: Good catch! It's been a while since I wrote the PR.

**Comment by haata** [comment]
Thanks!

**Comment by ryan-summers** [comment]
No problem - thanks for the help! :)

**Pull request merged**